### PR TITLE
[SPARK-18755][WIP][ML] Add Randomized Grid Search to Spark ML

### PR DESCRIPTION
**Pull request by tengpeng**
## What changes were proposed in this pull request?

Python sklearn has a randomized grid search for reducing the time for parameter tuning: 
1. If the candidate parameter values are all discrete, sampling with replacement.
2. If at least one candidate parameter is continuous, sampling without replacement.

This patch mimic the behavior of case 1 only. If we want to do 2, we need significant changes in `ParamGridBuilder` and other cross validation components. This requires more discussions. Thoughts?

Reference: http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html

## How was this patch tested?
Existing test + a new unit test.

### CODE REVIEW

**Comment by srowen** [comment]
I am not sure this approach adds much. Just lets you shuffle and sample the existing grid. Random hyperparam search is really about sampling a distribution for each hyperparam.

**Comment by tengpeng** [comment]
@srowen You are absolutely right. That's the what 2 aims to accomplish. I believe implementing 1 & 2 is the goal, like what they did in sklearn. Need some discussions on design.
